File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        if(head == NULL) return NULL;
        int size = 0;
        ListNode * temp = head;
        while(temp){
            size++;
            temp= temp->next;
        }
        int mid = size/2;
        if(mid < 1)  return NULL;
        int i = 0; 
        temp = head;
        ListNode * pre = NULL;
        while(i < mid && temp != NULL){
            i++;
            pre = temp;
            temp = temp->next;
        }
        // cout <<  pre->val <<  endl;
        if(temp != NULL){
            pre->next = temp->next;
            temp->next == NULL;
        }
        return head;
    }
};

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        if(head == null || head.next == null) return null;
        if(head.next.next == null) {
            head.next = null;
            return  head;
        }
       ListNode pre = null;
       ListNode cur = head;
       ListNode fast = head;
       while(fast != null && fast.next != null && fast.next.next != null){
            pre = cur;
            cur = cur.next;
            
            if(fast.next.next == null){
              fast = fast.next;
            } else {
                fast = fast.next.next;
            }
            // System.out.println(cur.val + " " + fast.val);
       }
       if(fast.next == null){
            pre.next = cur.next;
            cur.next = null;
       }
       else if(fast.next.next == null && fast.next != null){
            pre = cur;
            cur = cur.next; 
            pre.next = cur.next;
            cur.next = null;
       }

       return head;
    }
}